### ISSUE TITLE
Enum.value type annotation is ignored with custom `__new__` constructor

### ISSUE THREAD

**Issue by moosbruggerj** [microsoft/pyright#5741]
**Describe the bug**
When a custom `__new__` constructor for an enum is used, the `value` property always returns `Any`, even if a type stub for `value` is present.

**Code or Screenshots**
```python
from enum import Enum
from typing import TYPE_CHECKING, Literal
from typing_extensions import reveal_type

class Plain(Enum):
    MEMBER = 1

class Simple(bytes, Enum):
    MEMBER = (1, "description")
    description: str

    def __new__(cls, value: int, description: str):
        obj = bytes.__new__(cls, [value])
        obj._value_ = value
        obj.description = description
        return obj

class Annotated(bytes, Enum):
    MEMBER = (1, "description")
    description: str

    def __new__(cls, value: int, description: str):
        obj = bytes.__new__(cls, [value])
        obj._value_ = value
        obj.description = description
        return obj

    if TYPE_CHECKING:
        @property
        def value(self) -> Literal[1]: ...

reveal_type(Plain.MEMBER.value)
plain_value = Plain.MEMBER.value
reveal_type(plain_value)

reveal_type(Simple.MEMBER.value)
simple_value = Simple.MEMBER.value
reveal_type(simple_value)

reveal_type(Annotated.MEMBER.value)
value = Annotated.MEMBER.value
reveal_type(value)
```
As expected, the `Plain` enum has the correct literal type, but I would expect `Annotated` to have the correct type as well.
```
$ npx pyright@1.1.323 enum_test.py
enum_test.py:32:13 - information: Type of "Plain.MEMBER.value" is "Literal[1]"
enum_test.py:34:13 - information: Type of "plain_value" is "Literal[1]"
enum_test.py:36:13 - information: Type of "Simple.MEMBER.value" is "Any"
enum_test.py:38:13 - information: Type of "simple_value" is "Any"
enum_test.py:40:13 - information: Type of "Annotated.MEMBER.value" is "Any"
enum_test.py:42:13 - information: Type of "value" is "Any"
```
However, the `value` property annotation for both `Simple` and `Annotated` is as expected, when queried by the lsp on hover or code completion. (In my case by the neovim lsp)
```
# Simple
(property) value: (self: Self@Enum) -> Any

# Annotated
(property) value: (self: Self@Annotated) -> Literal[1]
```
The lsp server runs the same pyright version 1.1.323

**VS Code extension or command-line**
pyright 1.1.323 (cmd/lsp)

**Additional Context**

mypy 1.5.1 infers the type as expected:
```
enum_test.py:32: note: Revealed type is "Literal[1]?"
enum_test.py:34: note: Revealed type is "builtins.int"
enum_test.py:36: note: Revealed type is "Any"
enum_test.py:38: note: Revealed type is "Any"
enum_test.py:40: note: Revealed type is "Literal[1]"
enum_test.py:42: note: Revealed type is "Literal[1]"
```

**Comment by erictraut** [comment]
Hmm, the code in your code sample is questionable. The definition it provides for `value` under a `TYPE_CHECKING` conditional wouldn't work at runtime. It looks like you're relying on an undocumented (and probably unintended) behavior of mypy. I don't see anywhere in the [mypy documentation for enums](https://mypy.readthedocs.io/en/stable/literal_types.html#enums) that indicates that a dummy `value` property should indicate the intended type of the enum value. There's likewise nothing in the typing PEPs or Python documentation that indicates that a dummy `value` property should be interpreted in this way by a type checker.

**Comment by erictraut** [comment]
It was trivially easy for me to change pyright's logic to match mypy's behavior in this case, and I don't see any significant downside, so I've gone ahead and made that change. However, I'm not making any guarantees that this behavior won't change in the future. As I said above, I don't think this mechanism is documented anywhere, and I suspect it happens to work only by chance in mypy's implementation.

**Comment by moosbruggerj** [comment]
I copied this dummy style from pydantic (https://docs.pydantic.dev/latest/api/root_model/#pydantic.root_model.RootModel.model_dump) where it is used and seems to work for normal classes. :shrug:

**Comment by erictraut** [comment]
This is included in pyright 1.1.324, which I just published. It will also be included in a future release of pylance.
